esmvalcore/preprocessor/_time.py

```diff
@@ -1119,9 +1112,17 @@ def resample_hours(cube, interval, offset=0):
     if cube_period.total_seconds() / 3600 >= interval:
         raise ValueError(f"Data period ({cube_period}) should be lower than "
                          f"the interval ({interval})")
-    hours = range(0 + offset, 24, interval)
-    select_hours = iris.Constraint(time=lambda cell: cell.point.hour in hours)
-    return cube.extract(select_hours)
+    hours = [PartialDateTime(hour=h) for h in range(0 + offset, 24, interval)]
+    dates = time.units.num2date(time.points)
+    select = np.zeros(len(dates), dtype=bool)
+    for hour in hours:
+        select |= dates == hour
+    cube = _select_timeslice(cube, select)
```

Comment:
I think `select_timeslice` fails if time is a scalar dimension due to this line:

```py
time_dim = cube.coord_dims(coord)[0]
```

Do we need to consider this here? In `_extract_datetime` you considered this case explicitly.

Reply:
I can make it a bit nicer still, but using this function on a scalar cube really doesn't make sense: at best it will return the original cube.

Reply:
I am not saying we need to support this, but it would probably be nice to check if the coordinate is scalar and return a nice error message. Otherwise the line `time_dim = cube.coord_dims(coord)[0]` will probably produce an ugly `IndexError`.